winch/codegen/src/isa/aarch64/masm.rs

```diff
@@ -402,8 +448,17 @@ impl Masm for MacroAssembler {
         size: OperandSize,
     ) -> Result<()> {
         let (base, offset) = src.unwrap_offset();
-        self.asm
-            .add_ir(u64::try_from(offset).unwrap(), base, dst, size);
+        match Imm12::maybe_from_u64(offset as u64) {
+            Some(imm12) => self.asm.add_ir(imm12, base, dst, size),
+            None => {
+                self.with_scratch::<IntScratch, _>(|masm, scratch| {
+                    masm.asm
+                        .mov_ir(scratch.writable(), I::i64(offset), OperandSize::S64);
+                    masm.asm.add_rrr(scratch.inner(), base, dst, size);
+                    Ok(())
+                })?;
+            }
+        }
```

Comment:
Would it make sense to have `add_ir` methods on `masm` which encapsulate this pattern? (it looks similar below in `add` for example)